extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheResultInterceptor.java

```diff
@@ -59,8 +60,15 @@ public Object apply(Object k) {
                 }
             }
 
-        } catch (CacheException e) {
-            if (e.getCause() instanceof Exception) {
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof CacheException) {
+                CacheException cacheException = (CacheException) e.getCause();
+                if (cacheException.getCause() instanceof Exception) {
```

Comment:
Not sure I understand this particular one?

It looks a bit like dark magic, we should at least add comments about what we're doing here.

Reply:
It's not as dark as it seems :)

Exceptions happening during a cache computation are thrown from a `CompletableFuture` execution and are therefore encapsulated into an `ExecutionException` automatically. Since we use the functional interface `Function<Object, Object>` to pass the value loader to the cache, we also need to encapsulate any exception thrown during a cache computation into a `CacheException extends RuntimeException` because `Exception` cannot be thrown from a functional interface.

So if an `IOException` is thrown during a cache computation, it is encapsulated into a `CacheException` because of the functional interface used for the computation and again the `CacheException` is encapsulated into an `ExecutionException` because of the `CompletableFuture`.

This is the most likely scenario, but there are other possibilities as well (like the JVM throwing an Error) which are taken care of with a bunch of `else` blocks.

Reply:
I will add comments btw.

Reply:
@gsmet The exception handling code is now commented.